### PR TITLE
engine: optimize MVCCClearTimeRange

### PR DESCRIPTION
**Pull request by dt**
This is change is purely for optimization.

One common expected use-case of MVCCClearTimeRange is to revert
bulk-ingestion that often may create entire ranges of new keys, thus
meaning the entire range would meet the time predicate and be cleared.
Clearing all of those keys individually via Clear would result in a much
more expensive batch than a single ClearRange, so when iterating, this
change switches to buffer keys determined to match the time-bounds until
some threshold is met, and beyond that simply iterate tracking the
beginning of the span until the first non-matching key is observed, then
flush a single ClearRange for the matching span. If a non-matching key
is seen before that threshold is met, the buffered matching keys are
instead flushed as individual Clears.

This also changes the iteration to use an actual Iterator instead of
batch.Iterate, which allows using UnsafeKey to check the time bounds and
avoid allocating when iterating over non-matching keys. This to is
expected in many bulk-ingestions -- just as clustered data might result
in a swath of all matching keys, it also likely results in large swaths
of all non-matching keys, so avoiding extra expense in skipping
non-matching keys is also expected to be a win in those cases.

Finally, to test these cases that only come up when dealing with larger
numbers of keys, this change also adds some random-data based tests to
complement the existing hand-made test cases that remain small enough
and simple enough for a human to trace their expected results.

Release note: none.

### CODE REVIEW

**Review by thoszhang — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/39050)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dt, @lucy-zhang, and @nvanbenschoten)

___
*[pkg/storage/engine/mvcc.go, line 1860 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/39050#-LkUpAtZBaC2xELcOsCD:-LkUpAtZBaC2xELcOsCE:b-hvjmk7) ([raw file](https://github.com/cockroachdb/cockroach/blob/d03e7278db9adb415f0aef46db5f380189b2a6ab/pkg/storage/engine/mvcc.go#L1860)):*
> ```Go
> ) error {
> 
> 	const useRangeClearThreshold = 64
> ```

Is there a reason for this value?

**Review by dt — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/39050)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @lucy-zhang and @nvanbenschoten)

___
*[pkg/storage/engine/mvcc.go, line 1860 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/39050#-LkUpAtZBaC2xELcOsCD:-LkUyS_l2PqYKKCh-cas:b3zeyk7) ([raw file](https://github.com/cockroachdb/cockroach/blob/d03e7278db9adb415f0aef46db5f380189b2a6ab/pkg/storage/engine/mvcc.go#L1860)):*
<details><summary><i>Previously, lucy-zhang (Lucy Zhang) wrote…</i></summary><blockquote>

Is there a reason for this value?
</blockquote></details>

I asked #core and that is what they said, though they also said it was pulled from thin air.

**Comment by dt** [comment]
I have pagination implemented locally on top of this change, but I think I'll hold off and send it separately after this merges.

**Review by nvb — COMMENTED** [review]
> Reviewed 3 of 3 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/39050)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dt, @lucy-zhang, and @nvanbenschoten)

___
*[pkg/storage/engine/mvcc.go, line 1860 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/39050#-LkUpAtZBaC2xELcOsCD:-LkeJY19BrG-4gKPxgkw:byfxrjx) ([raw file](https://github.com/cockroachdb/cockroach/blob/d03e7278db9adb415f0aef46db5f380189b2a6ab/pkg/storage/engine/mvcc.go#L1860)):*
<details><summary><i>Previously, dt (David Taylor) wrote…</i></summary><blockquote>

I asked #core and that is what they said, though they also said it was pulled from thin air.
</blockquote></details>

Could you add a comment referencing `clearRangeData`?
___
*[pkg/storage/engine/mvcc.go, line 1862 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/39050#-LkeJwcI3621OZReqr4S:-LkeJwcI3621OZReqr4T:b-2qo2nn) ([raw file](https://github.com/cockroachdb/cockroach/blob/d03e7278db9adb415f0aef46db5f380189b2a6ab/pkg/storage/engine/mvcc.go#L1862)):*
> ```Go
> 	const useRangeClearThreshold = 64
> 
> 	// When iterating, instead of immediately clearing a matching key, instead we
> ```

instead ... instead
___
*[pkg/storage/engine/mvcc.go, line 1871 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/39050#-LkeJsuP-n9NQBqNPOcx:-LkeJsuP-n9NQBqNPOcy:b-ek0wiy) ([raw file](https://github.com/cockroachdb/cockroach/blob/d03e7278db9adb415f0aef46db5f380189b2a6ab/pkg/storage/engine/mvcc.go#L1871)):*
> ```Go
> 	// entire span may likely match and thus could be cleared in one ClearRange
> 	// instead of hundreds of thousands of individual Clears.
> 	buf := make([]MVCCKey, 0, 100)
> ```

Should the capacity for this be `useRangeClearThreshold`?

EDIT: addressed in next commit.
___
*[pkg/storage/engine/mvcc.go, line 1887 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/39050#-LkeK_EQ-J5lEGBItV0E:-LkeK_ERERbyeEhvb9iY:b-lfgbz3) ([raw file](https://github.com/cockroachdb/cockroach/blob/d03e7278db9adb415f0aef46db5f380189b2a6ab/pkg/storage/engine/mvcc.go#L1887)):*
> ```Go
> 	it := batch.NewIterator(IterOptions{LowerBound: key, UpperBound: endKey})
> 	defer it.Close()
> 	it.Seek(MVCCKey{Key: key})
> ```

nit: usually the `it.Seek` is included as part of the `for` loop like:
```
for it.Seek(MVCCKey{Key: key}); ; it.Next() {
```
___
*[pkg/storage/engine/mvcc.go, line 1901 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/39050#-LkeKxPh2GTmp99ukYnK:-LkeKxPi3JTNJi-eHplV:b-qa4mu3) ([raw file](https://github.com/cockroachdb/cockroach/blob/d03e7278db9adb415f0aef46db5f380189b2a6ab/pkg/storage/engine/mvcc.go#L1901)):*
> ```Go
> 		}
> 
> 		// We need to check for and fail on any intents in out time-range as we do
> ```

s/out/our/
___
*[pkg/storage/engine/mvcc.go, line 1921 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/39050#-LkeLQ736LGt6TOyEz3b:-LkeLQ744JSWK3tX5Zv5:b88sjeh) ([raw file](https://github.com/cockroachdb/cockroach/blob/d03e7278db9adb415f0aef46db5f380189b2a6ab/pkg/storage/engine/mvcc.go#L1921)):*
> ```Go
> 
> 		if startTime.Less(k.Timestamp) && !endTime.Less(k.Timestamp) {
> 			if len(rangeStart.Key) == 0 {
> ```

What will happen if we pass `roachpb.KeyMin` as the starting key to this function?
___
*[pkg/storage/engine/mvcc.go, line 1921 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/39050#-LkeLYla0tFFzIJtP2Is:-LkeLYla0tFFzIJtP2It:b3i537y) ([raw file](https://github.com/cockroachdb/cockroach/blob/d03e7278db9adb415f0aef46db5f380189b2a6ab/pkg/storage/engine/mvcc.go#L1921)):*
> ```Go
> 
> 		if startTime.Less(k.Timestamp) && !endTime.Less(k.Timestamp) {
> 			if len(rangeStart.Key) == 0 {
> ```

I think this would be clearer and would avoid code duplication if you created two closures at the top of this function right below `buf` and `rangeStart`:
```
addToBatch := func(MVCCKey) { ... }
flushBatch := func(MVCCKey) { ... }
```

**Review by dt — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/39050)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @lucy-zhang and @nvanbenschoten)

___
*[pkg/storage/engine/mvcc.go, line 1860 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/39050#-LkUpAtZBaC2xELcOsCD:-LkehFkk5uLspD2LLF00:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d03e7278db9adb415f0aef46db5f380189b2a6ab/pkg/storage/engine/mvcc.go#L1860)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Could you add a comment referencing `clearRangeData`?
</blockquote></details>

Done.
___
*[pkg/storage/engine/mvcc.go, line 1862 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/39050#-LkeJwcI3621OZReqr4S:-LkehKVS01sFDxSxjL7n:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d03e7278db9adb415f0aef46db5f380189b2a6ab/pkg/storage/engine/mvcc.go#L1862)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

instead ... instead
</blockquote></details>

Done.
___
*[pkg/storage/engine/mvcc.go, line 1871 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/39050#-LkeJsuP-n9NQBqNPOcx:-LkehVkZ9fJU6Zxwtn1l:b78viaz) ([raw file](https://github.com/cockroachdb/cockroach/blob/d03e7278db9adb415f0aef46db5f380189b2a6ab/pkg/storage/engine/mvcc.go#L1871)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Should the capacity for this be `useRangeClearThreshold`?

EDIT: addressed in next commit.
</blockquote></details>

squashed.
___
*[pkg/storage/engine/mvcc.go, line 1887 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/39050#-LkeK_EQ-J5lEGBItV0E:-LkehT19CJHVNk2Q4wWu:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d03e7278db9adb415f0aef46db5f380189b2a6ab/pkg/storage/engine/mvcc.go#L1887)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

nit: usually the `it.Seek` is included as part of the `for` loop like:
```
for it.Seek(MVCCKey{Key: key}); ; it.Next() {
```
</blockquote></details>

Done.
___
*[pkg/storage/engine/mvcc.go, line 1901 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/39050#-LkeKxPh2GTmp99ukYnK:-LkehSE18bmT3MBBLJkw:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d03e7278db9adb415f0aef46db5f380189b2a6ab/pkg/storage/engine/mvcc.go#L1901)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

s/out/our/
</blockquote></details>

Done.
___
*[pkg/storage/engine/mvcc.go, line 1921 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/39050#-LkeLYla0tFFzIJtP2Is:-Lkeg83kETLb5D5s61U5:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d03e7278db9adb415f0aef46db5f380189b2a6ab/pkg/storage/engine/mvcc.go#L1921)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

I think this would be clearer and would avoid code duplication if you created two closures at the top of this function right below `buf` and `rangeStart`:
```
addToBatch := func(MVCCKey) { ... }
flushBatch := func(MVCCKey) { ... }
```
</blockquote></details>

Done.
___
*[pkg/storage/engine/mvcc.go, line 1921 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/39050#-LkeLQ736LGt6TOyEz3b:-LkegLm8ECQ4gZgL-R22:bdeo7op) ([raw file](https://github.com/cockroachdb/cockroach/blob/d03e7278db9adb415f0aef46db5f380189b2a6ab/pkg/storage/engine/mvcc.go#L1921)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

What will happen if we pass `roachpb.KeyMin` as the starting key to this function?
</blockquote></details>

There can't actually be a real key with that key, can there? `rangeStart` is set to the first _matching_ key, so we'd need to read a 0 byte key from the iterator.

**Review by nvb — APPROVED** [review]
:lgtm:

> Reviewed 1 of 1 files at r3.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/39050#-:-Lkfxb8O5u03xoAPgJdv:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @dt and @lucy-zhang)

___
*[pkg/storage/engine/mvcc.go, line 1921 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/39050#-LkeLQ736LGt6TOyEz3b:-LkfxNPFCOuiT6Vy6XHI:bdcv6l8) ([raw file](https://github.com/cockroachdb/cockroach/blob/d03e7278db9adb415f0aef46db5f380189b2a6ab/pkg/storage/engine/mvcc.go#L1921)):*
<details><summary><i>Previously, dt (David Taylor) wrote…</i></summary><blockquote>

There can't actually be a real key with that key, can there? `rangeStart` is set to the first _matching_ key, so we'd need to read a 0 byte key from the iterator.
</blockquote></details>

Good point.
___
*[pkg/storage/engine/mvcc.go, line 1893 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/39050#-LkfxZbc00d1dbUaC7yX:-LkfxZbc00d1dbUaC7yY:b-y9etto) ([raw file](https://github.com/cockroachdb/cockroach/blob/35d9441d91c4106773ff47dc3f95344f342e2c4a/pkg/storage/engine/mvcc.go#L1893)):*
> ```Go
> 	}
> 
> 	flushClearedKeys := func(k MVCCKey) error {
> ```

Can we make it a little more clear what the argument is here?

**Review by dt — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/39050)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @dt, @lucy-zhang, and @nvanbenschoten)

___
*[pkg/storage/engine/mvcc.go, line 1893 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/39050#-LkfxZbc00d1dbUaC7yX:-LkgA0nv5YGCh6FcLefh:b-9etxcq) ([raw file](https://github.com/cockroachdb/cockroach/blob/35d9441d91c4106773ff47dc3f95344f342e2c4a/pkg/storage/engine/mvcc.go#L1893)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Can we make it a little more clear what the argument is here?
</blockquote></details>

Renamed it to `nonMatching` to help tie it to the big comment above talking about when we have to flush.

**Pull request merged**
